File: akka-docs/src/test/scala/docs/stream/operators/sourceorflow/MapWithResource.scala
/*
 * Copyright (C) 2022 Lightbend Inc. <https://www.lightbend.com>
 */

package docs.stream.operators.sourceorflow
import akka.actor.ActorSystem
import akka.stream.scaladsl.Source

import java.net.URL

object MapWithResource {
  implicit val actorSystem: ActorSystem = ???

  //#mapWithResource-blocking-api
  trait DBDriver {
    def create(url: URL, userName: String, password: String): Connection
  }
  trait Connection {
    def close(): Unit
  }
  trait Database {
    //blocking query
    def doQuery(connection: Connection, query: String): QueryResult = ???
  }
  trait QueryResult {
    def hasMore: Boolean
    // potentially blocking retrieval of each element
    def next(): DataBaseRecord
    // potentially blocking retrieval all element
    def toList(): List[DataBaseRecord]
  }
  trait DataBaseRecord
  //#mapWithResource-blocking-api
  val url: URL = ???
  val userName = "Akka"
  val password = "Hakking"
  val dbDriver: DBDriver = ???
  def mapWithResourceExample(): Unit = {
    //#mapWithResource
    //some database for JVM
    val db: Database = ???
    Source(
      List(
        "SELECT * FROM article-0000 ORDER BY gmtModified DESC LIMIT 100;",
        "SELECT * FROM article-0001 ORDER BY gmtModified DESC LIMIT 100;"))
      .mapWithResource(() => dbDriver.create(url, userName, password))(
        (connection, query) => db.doQuery(connection, query).toList(),
        conn => {
          conn.close()
          None
        })
      .mapConcat(identity)
      .runForeach(println)
    //#mapWithResource
  }
}

File: akka-docs/src/test/java/jdocs/stream/operators/sourceorflow/MapWithResource.java
/*
 * Copyright (C) 2022-2025 Lightbend Inc. <https://www.lightbend.com>
 */

package jdocs.stream.operators.sourceorflow;

import akka.actor.ActorSystem;
import akka.stream.javadsl.Source;
import java.net.URL;
import java.util.Arrays;
import java.util.List;
import java.util.Optional;

public interface MapWithResource {
  // #mapWithResource-blocking-api
  interface DBDriver {
    Connection create(URL url, String userName, String password);
  }

  interface Connection {
    void close();
  }

  interface Database {
    // blocking query
    QueryResult doQuery(Connection connection, String query);
  }

  interface QueryResult {
    boolean hasMore();

    // potentially blocking retrieval of each element
    DatabaseRecord next();

    // potentially blocking retrieval all element
    List<DatabaseRecord> toList();
  }

  interface DatabaseRecord {}
  // #mapWithResource-blocking-api

  default void mapWithResourceExample() {
    final ActorSystem system = null;
    final URL url = null;
    final String userName = "Akka";
    final String password = "Hakking";
    final DBDriver dbDriver = null;
    // #mapWithResource
    // some database for JVM
    final Database db = null;
    Source.from(
            Arrays.asList(
                "SELECT * FROM shop ORDER BY article-0000 order by gmtModified desc limit 100;",
                "SELECT * FROM shop ORDER BY article-0001 order by gmtModified desc limit 100;"))
        .mapWithResource(
            () -> dbDriver.create(url, userName, password),
            (connection, query) -> db.doQuery(connection, query).toList(),
            connection -> {
              connection.close();
              return Optional.empty();
            })
        .mapConcat(elems -> elems)
        .runForeach(System.out::println, system);
    // #mapWithResource
  }
}
